### PR TITLE
Should finish race effect on END

### PR DESCRIPTION
**Pull request by restrry**
a breaking change. 
race should be finished if any of effects resolved with END (by analogy with `all`)
https://github.com/redux-saga/redux-saga/issues/1362
https://github.com/redux-saga/redux-saga/issues/1354

### CODE REVIEW

File: packages/core/test/interpreter/race.js [hunk 1]
```diff
@@ -73,33 +73,33 @@ test('saga race between array of effects handling', assert => {
...
-    actual.push(
+    try {
       yield io.race({
         event: io.take('action'),
         timeout: timeout.promise,
-      }),
-    )
+      })
+    } finally {
+      cancelled = true
```
Comment (shinima) [inline packages/core/test/interpreter/race.js]: It seems incorrect to do assignment in finally block. `cancelled = true` will always executes here regardless of the race result.

Comment (restrry) [reply packages/core/test/interpreter/race.js]: oops. forgot to remove delay.

Comment (restrry) [reply packages/core/test/interpreter/race.js]: I aligned the test with happy path version to  emphasise the difference between them
https://github.com/redux-saga/redux-saga/blob/af4b1b41f6f336bb4668fe2be569819a5ca3ee62/packages/core/test/interpreter/race.js#L8

**Comment by shinima** [comment]
Changes looks good to me. But I'm worrying about the cases when we place takeMaybe in race/all.

```js
function* gen() {
   const allResult = yield all([
    takeMaybe('some-action'), // will take an END
    somePromise,
  ])
  expect.deepEqual(allResult, [END, resolvedValueOfPromise])
}
```

When `takeMay` take an END from the channel, END is passed to `all` and then `all` terminates the task. So we could never take an END using takeMay if takeMay is nested in all/race.

I think maybe we should treat END as a normal result in all/race. How do you think ? Point out my mistakes if I'm wrong.

**Comment by shinima** [comment]
I may not explain myself clearly... Actually I'm talking about `takeMaybe` in `all`. In current implementation, users could not use `takeMaybe` to take an END if `takeMaybe` is in `all`.

It my fault that `takeMaybe` is not related to this PR. I'll make another PR when this PR gets merged. This PR itself is very good that it makes `all` and `race` more consistent when dealing with END.

**Comment by Andarist** [comment]
At least it's consistent with `all` now 👍 Thanks you both for great work, I'll try to catch up with other issues/PRs in following days

**Comment by Andarist** [comment]
Damn, I cant remember what cases I've wanted to check 🤦‍♂️

**Pull request merged**